### PR TITLE
32851: Fix mypy errors for azure-maps-render

### PR DESCRIPTION
**Pull request by Lakicar95**
# Description

This change will fix mypy typing errors for azure-maps-render.

Current mypy version in tox.ini: `mypy_version=1.6.1`

Resolves: https://github.com/Azure/azure-sdk-for-python/issues/32851

# How it is tested?
`azure-sdk-for-python\sdk\maps\azure-maps-render> tox run -e mypy -c ../../../eng/tox/tox.ini --root . `

### CODE REVIEW

File: sdk/eventgrid/azure-eventgrid/samples/async_samples/sample_authentication_async.py [hunk 1]
```diff
@@ -20,8 +20,8 @@
 from azure.eventgrid.aio import EventGridPublisherClient
 from azure.core.credentials import AzureKeyCredential
 
-topic_key = os.environ["EVENTGRID_TOPIC_KEY"]
-endpoint = os.environ["EVENTGRID_TOPIC_ENDPOINT"]
+topic_key = os.environ["EVENTGRID_TOPIC_KEY"] or "your-eventgrid-topic-key"
+endpoint = os.environ["EVENTGRID_TOPIC_ENDPOINT"] or "your-eventgrid-topic-endpoint"
```
Comment (kristapratico) [inline sdk/eventgrid/azure-eventgrid/samples/async_samples/sample_authentication_async.py:24]: We generally try to keep a PR focused to one library at a time. Can we revert these changes for eventgrid?

File: sdk/maps/azure-maps-render/azure/maps/render/_render_client.py [hunk 2]
```diff
@@ -441,12 +441,12 @@ def get_copyright_from_bounding_box(
...
-        return self._render_client.get_copyright_from_bounding_box(
-            south_west=(bounding_box.south,bounding_box.west),
-            north_east=(bounding_box.north,bounding_box.east),
+        return Copyright(**self._render_client.get_copyright_from_bounding_box(
+            south_west=[bounding_box.south,bounding_box.west],
+            north_east=[bounding_box.north,bounding_box.east],
             include_text= "yes" if _include_text else "no",
             **kwargs
-        )
+        ).__dict__)
```
Comment (kristapratico) [inline sdk/maps/azure-maps-render/azure/maps/render/_render_client.py:449]: What was the mypy error here?

Comment (Lakicar95) [reply sdk/maps/azure-maps-render/azure/maps/render/_render_client.py:449]: Expected return type is from: `sdk/maps/azure-maps-render/azure/maps/render/models/_models.py`
But the one `self._render_client.get_copyright_from_bounding_box` is returning is from the _generated dir: `sdk/maps/azure-maps-render/azure/maps/render/_generated/models/_models.py`. We have to convert generated type to the custom one.

File: sdk/eventgrid/azure-eventgrid/samples/async_samples/sample_authentication_async.py (hunk 1 above)
Comment (Lakicar95) [reply sdk/eventgrid/azure-eventgrid/samples/async_samples/sample_authentication_async.py]: Done

File: sdk/maps/azure-maps-render/azure/maps/render/_render_client.py (hunk 2 above)
Comment (kristapratico) [reply sdk/maps/azure-maps-render/azure/maps/render/_render_client.py:449]: Ah good point. However, it seems that the tests and samples for this library do expect `Copyright` in the shape of the generated model. I don't have context for whether this was intended to be transformed to the custom type (and overlooked), or if we decided the generated type was acceptable. Let me see if I can tag the library owner to weigh in.

Comment (kristapratico) [reply sdk/maps/azure-maps-render/azure/maps/render/_render_client.py:449]: @alextts627 - can you help answer this question?

Comment (Lakicar95) [reply sdk/maps/azure-maps-render/azure/maps/render/_render_client.py:449]: That's the way these things are handled within search, route, geolocation.. But let's double check :)

One more thing... I need some help regarding autorest. I wasn't able to generate py3 files in models dir: sdk/maps/azure-maps-search/azure/maps/search/_generated/models/_models_py3.py

I was using this swagger file as input: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/maps/data-plane/Search/preview/1.0/search.json

Using legacy msrest mode I was able to generate just: _models.py, but not _models_py3.py.

Note: I fixed issue in the doc, by replacing: `["SharedKey"]["in"] = "header"`

Can you help me with concrete parameters for generating code.. This existing README.md is not up to date. I also tried with cloning azure-rest-api-specs repo and couldn't produce _models_py3.py file.

Thanks :)

Comment (kristapratico) [reply sdk/maps/azure-maps-render/azure/maps/render/_render_client.py:449]: Just for clarity, the reason I want to double-check with the library owner on this is because from generated type --> custom type, it's unclear to me whether the intention there was to rename the `regions` property to `regional_copyrights`. That's not how things will deserialize today, so we'll need to update the code if that was the intention (and document the breaking change).

Regarding autorest - that's by design. We no longer generate with a _models_py3 file, that was only necessary during the time that the SDK still supported python 2.7 (py2 and py3 models used to be separated). I'm guessing it's been awhile since this library has been regenerated and so it still had the legacy _models_py3 file.

Why do you believe the existing readme to not be up to date? From what I can tell, the tag in the README matches the default API version for this version of the library.

Comment (Lakicar95) [reply sdk/maps/azure-maps-render/azure/maps/render/_render_client.py:449]: @kristapratico thanks a lot for explanation

Ok, now I understand why I couldn't generate _models_py3.py file. But, hm.. reverse search uses models from _models_py3.py to validate response (which causes this error by the way: https://github.com/Azure/azure-sdk-for-python/issues/32565.. I created pr for that which we can merge as this py3 is still used, but not generated with autorest: https://github.com/Azure/azure-sdk-for-python/pull/34977)? 

The other thing is why README.md seems outdated to me is that I couldn't generate _models.py (regular not py3) without using msrest mode (which is legacy and it is removed: https://github.com/Azure/azure-rest-api-specs/pull/28269). Is there any other way to modify README.md for azure-maps-search to generate _models.py ?

Comment (kristapratico) [reply sdk/maps/azure-maps-render/azure/maps/render/_render_client.py:449]: > Ok, now I understand why I couldn't generate _models_py3.py file. But, hm.. reverse search uses models from _models_py3.py to validate response (which causes this error by the way: https://github.com/Azure/azure-sdk-for-python/issues/32565.. I created pr for that which we can merge as this py3 is still used, but not generated with autorest: https://github.com/Azure/azure-sdk-for-python/pull/34977)?

In this case, the spec needs to be updated to match the service response. The py3 file will be deleted upon next generation so the update won't persist if we change it manually. Additionally, since the spec file is used to generate more languages than just Python, fixing this in the spec will propagate to all the other libraries.

> The other thing is why README.md seems outdated to me is that I couldn't generate _models.py (regular not py3) without using msrest mode (which is legacy and it is removed: https://github.com/Azure/azure-rest-api-specs/pull/28269). Is there any other way to modify README.md for azure-maps-search to generate _models.py ?

Since this library uses swagger-based code generation, `msrest` mode has to be used if models are needed. The Python SDK is transitioning from generation from swagger definitions to [typespec](https://typespec.io/) and the service definition for maps has not be converted over yet. All new services / versions are defined in typespec, but some of the existing / not recently updated services haven't made the jump yet.

**Comment by dubiety** [comment]
Hi @kristapratico , [alextts627](https://github.com/alextts627) is no longer in the team. I'll be responsible to discuss the fix with @Lakicar95 .

File: sdk/maps/azure-maps-render/azure/maps/render/_render_client.py (hunk 2 above)
Comment (Lakicar95) [reply sdk/maps/azure-maps-render/azure/maps/render/_render_client.py:449]: @kristapratico Thanks, just regarding specs, it's already updated: https://github.com/Azure/azure-rest-api-specs/pull/17151

Comment (kristapratico) [reply sdk/maps/azure-maps-render/azure/maps/render/_render_client.py:449]: > @kristapratico Thanks, just regarding specs, it's already updated: [Azure/azure-rest-api-specs#17151](https://github.com/Azure/azure-rest-api-specs/pull/17151)

Great, thanks for sharing! In that case, is it possible to update https://github.com/Azure/azure-sdk-for-python/pull/34977 with the latest code generation? For anywhere the code uses models_py3, can we update references to use _models.py?

Comment (Lakicar95) [reply sdk/maps/azure-maps-render/azure/maps/render/_render_client.py:449]: Upon trying to regenerate new code described in this draft pr: https://github.com/Azure/azure-sdk-for-python/pull/35029 I managed to resolve the issue (it will replace usage of _models_py3.py with regular _models.py), but there is a lot of work to be done before fixing tests/everything.. Can you please take a brief look at that :)

Once again, thanks a lot for your help! :D

File: sdk/maps/azure-maps-render/samples/sample_get_map_tileset.py [hunk 3]
```diff
@@ -18,7 +18,7 @@
 
 import os
 
-subscription_key = os.getenv("AZURE_SUBSCRIPTION_KEY")
+subscription_key = os.getenv("AZURE_SUBSCRIPTION_KEY") or "your subscription key"
```
Comment (dubiety) [inline sdk/maps/azure-maps-render/samples/sample_get_map_tileset.py:21]: Will the below statement is better?
```python
subscription_key = os.getenv("AZURE_SUBSCRIPTION_KEY", "your subscription key")
```

**Review by dubiety — DISMISSED** [review]

**Comment by Lakicar95** [comment]
@dubiety I added new commits to address your comments.

File: sdk/maps/azure-maps-render/samples/async_samples/sample_authentication_async.py [hunk 4]
```diff
@@ -51,7 +51,7 @@ async def authentication_maps_service_client_with_aad_credential_async():
     from azure.maps.render.aio import MapsRenderClient
 
     credential = DefaultAzureCredential()
-    maps_client_id = os.getenv("AZURE_MAPS_CLIENT_ID")
```
Comment (dubiety) [inline sdk/maps/azure-maps-render/samples/async_samples/sample_authentication_async.py:54]: This line is not changed.

File: sdk/maps/azure-maps-render/samples/async_samples/sample_authentication_async.py [hunk 5]
```diff
@@ -51,7 +51,7 @@ async def authentication_maps_service_client_with_aad_credential_async():
...
+    maps_client_id = os.getenv("AZURE_MAPS_CLIENT_ID") or "your subscription key"
```
Comment (Lakicar95) [inline sdk/maps/azure-maps-render/samples/async_samples/sample_authentication_async.py:54]:
```suggestion
    maps_client_id = os.getenv("AZURE_MAPS_CLIENT_ID", "your client id")
```

File: sdk/maps/azure-maps-render/samples/async_samples/sample_authentication_async.py (hunk 4 above)
Comment (Lakicar95) [reply sdk/maps/azure-maps-render/samples/async_samples/sample_authentication_async.py:54]: done, thanks

**Comment by dubiety** [comment]
Hi @Lakicar95 looks like we have some [test failures](https://dev.azure.com/azure-sdk/public/_TestManagement/Runs?runId=47326818&_a=resultSummary&resultId=100000) due to the record test files not found. You can refer to the document here: https://github.com/Azure/azure-sdk-for-python/blob/main/doc/dev/tests.md#update-test-recordings for update test recordings. Let me know if you have any question, thanks a lot!
